### ISSUE TITLE
Rename Parser to DNMParser, create Parser protocol

### ISSUE THREAD

**Issue by jsbean** [jsbean/DNM#230]
```Swift
protocol Parser {
    var startLocation: Int { get set }
    var didFail: Bool { get set }
    func parseWith(scanner: NSScanner) -> Any?
    func resultWith(scanner: NSScanner) -> Any
    func unwindScannerIfNecessary(scanner: NSScanner)
    func setStartLocationWith(scanner: NSScanner)
}

extension Parser {
    func parseWith(scanner: NSScanner) -> Any? {
        setStartLocationWith(scanner: NSScanner)
        didFail = false // prove true
        defer { unwindScannerIfNecessary(scanner: NSScanner) }
        return resultWith(scanner: NSScanner) -> Any?
    }

    func setStartLocationWith(scanner: NSScanner) { ... }

    func configureScanLocationFor(scanner: NSScanner) {
        if didFail { scanner.scanLocation = startLocation }
    }
}
```

In `resultWith(scanner: NSScanner)`, set didFail if any subqueries fail.

**Comment by jsbean** [comment]
```Swift
DurationParser: Parser
DurationIntervalParser: Parser
DNMParser: Parser
IndentationParser: Parser
```

Usage:
```Swift
if let valueOfType = TypeParser().parseWith(scanner) as? Type {
    // success
}
```

**Comment by jsbean** [comment]
Possible to override return type?

**Comment by jsbean** [comment]
Closed with a3e4de5d6b7ecb6f12ba05d4d55e70f3769da7fe
